src/herder/test/UpgradesTests.cpp

```diff
@@ -1524,6 +1524,112 @@ TEST_CASE("upgrade to version 11", "[upgrades]")
     }
 }
 
+TEST_CASE("upgrade to version 12", "[upgrades]")
+{
+    VirtualClock clock;
+    auto cfg = getTestConfig();
+    cfg.USE_CONFIG_FOR_GENESIS = false;
+
+    auto app = createTestApplication(clock, cfg);
+    app->start();
+    executeUpgrade(*app, makeProtocolVersionUpgrade(11));
+
+    auto& lm = app->getLedgerManager();
+    uint32_t oldProto = 11;
+    uint32_t newProto = 12;
+    auto root = TestAccount{*app, txtest::getRoot(app->getNetworkID())};
+
+    for (size_t i = 0; i < 10; ++i)
+    {
+        auto stranger =
+            TestAccount{*app, txtest::getAccount(fmt::format("stranger{}", i))};
+        TxSetFramePtr txSet =
+            std::make_shared<TxSetFrame>(lm.getLastClosedLedgerHeader().hash);
+        uint32_t ledgerSeq = lm.getLastClosedLedgerNum() + 1;
+        uint64_t minBalance = lm.getLastMinBalance(5);
+        uint64_t big = minBalance + ledgerSeq;
+        uint64_t closeTime = 60 * 5 * ledgerSeq;
+        txSet->add(root.tx({txtest::createAccount(stranger, big)}));
+        // Provoke sortForHash and hash-caching:
+        txSet->getContentsHash();
+
+        // On 4th iteration of advance (a.k.a. ledgerSeq 5), perform a
+        // ledger-protocol version upgrade to the new protocol, to
+        // start new-style merges (no shadows)
+        auto upgrades = xdr::xvector<UpgradeType, 6>{};
+        if (ledgerSeq == 5)
+        {
+            auto ledgerUpgrade = LedgerUpgrade{LEDGER_UPGRADE_VERSION};
+            ledgerUpgrade.newLedgerVersion() = newProto;
+            auto v = xdr::xdr_to_opaque(ledgerUpgrade);
+            upgrades.push_back(UpgradeType{v.begin(), v.end()});
+            CLOG(INFO, "Ledger")
+                << "Ledger " << ledgerSeq << " upgrading to v" << newProto;
+        }
+        StellarValue sv(txSet->getContentsHash(), closeTime, upgrades,
+                        STELLAR_VALUE_BASIC);
+        lm.closeLedger(LedgerCloseData(ledgerSeq, txSet, sv));
+        auto& bm = app->getBucketManager();
+        auto mc = bm.readMergeCounters();
+        if (ledgerSeq < 5)
+        {
+            REQUIRE(mc.mPreShadowRemovalProtocolMerges != 0);
+        }
+        else
+        {
+            auto& lev0 = bm.getBucketList().getLevel(0);
+            auto& lev1 = bm.getBucketList().getLevel(1);
+            auto lev0Curr = lev0.getCurr();
+            auto lev0Snap = lev0.getSnap();
+            auto lev1Curr = lev1.getCurr();
+            auto lev1Snap = lev1.getSnap();
+            auto getVers = [](std::shared_ptr<Bucket> b) -> uint32_t {
+                return BucketInputIterator(b).getMetadata().ledgerVersion;
+            };
+            // post-shadow-removal merges
+            // We can't be quite sure how many in-progress merges actually
+            // finished (due to background thread races), so we check that the
+            // number of merges in between max possible (e.g., all kicked off
```

Comment:
This seems a little odd: can't we wait here until the merges complete, before continuing?

Reply:
fixed